src/core/expression/predicate.visitor.js

```diff
@@ -40,22 +40,24 @@ export class PredicateVisitor extends Visitor {
 		const assert = this.assertFactory(name);
 		const map = new Set();
 
-		let rCastAs;
+		let parse, rt;
 		if (isArray(r)) {
 			if (r.length) {
-				rCastAs = castAsFactory(r[0]);
+				rt = getType(r[0]);
+				parse = parseFactory(rt);
 				r.forEach(x => map.add('' + x));
 			} else {
-				rCastAs = identity;
+				parse = identity;
 			}
 		} else {
-			rCastAs = castAsFactory(r);
+			rt = getType(r);
+			parse = parseFactory(rt);
 		}
 
 		const equals = assert.equals;
```

Comment:
don't we need to parse `r`?

Reply:
I think it doesn't make sense, because we're just parsing `l` to type which `r` already has.